pkg/storage/secret/encryption/encrypted_value_store.go

```diff
@@ -206,3 +206,87 @@ func (s *encryptedValStorage) Delete(ctx context.Context, namespace, name string
 
 	return nil
 }
+
+func (s *encryptedValStorage) ListAll(ctx context.Context, opts contracts.ListOpts) ([]*contracts.EncryptedValue, error) {
+	ctx, span := s.tracer.Start(ctx, "EncryptedValueStorage.ListAll", trace.WithAttributes(
+		attribute.Int64("limit", opts.Limit),
+		attribute.Int64("offset", opts.Offset),
+	))
+	defer span.End()
+
+	req := listAllEncryptedValues{
+		SQLTemplate: sqltemplate.New(s.dialect),
+		Limit:       opts.Limit,
+		Offset:      opts.Offset,
+	}
+
+	query, err := sqltemplate.Execute(sqlEncryptedValueListAll, req)
+	if err != nil {
+		return nil, fmt.Errorf("execute template %q: %w", sqlEncryptedValueListAll.Name(), err)
+	}
+
+	rows, err := s.db.QueryContext(ctx, query, req.GetArgs()...)
+	if err != nil {
+		return nil, fmt.Errorf("listing encrypted values %q: %w", sqlEncryptedValueListAll.Name(), err)
+	}
+	defer func() { _ = rows.Close() }()
+
+	encryptedValues := make([]*contracts.EncryptedValue, 0)
+	for rows.Next() {
+		var row EncryptedValue
+		err = rows.Scan(
+			&row.Namespace,
+			&row.Name,
+			&row.Version,
+			&row.EncryptedData,
+			&row.Created,
+			&row.Updated,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error reading data key row: %w", err)
+		}
+
+		encryptedValues = append(encryptedValues, &contracts.EncryptedValue{
+			Namespace:     row.Namespace,
+			Name:          row.Name,
+			Version:       row.Version,
+			EncryptedData: row.EncryptedData,
+			Created:       row.Created,
+			Updated:       row.Updated,
+		})
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("read rows error: %w", err)
+	}
+
+	return encryptedValues, nil
+}
+
+func (s *encryptedValStorage) CountAll(ctx context.Context) (int64, error) {
```

Comment:
If I understood it properly, we'll use `CountAll` to paginate through the `ListAll`? Will the operation lock the encrypted value table? Otherwise there could be a problem:
- CountAll returns (`a1`, `b1`, `c1`) = 3
- Someone creates a new value with name `a2`
- ListAll will miss `c1` ?

Reply:
> Making it to lock will mean having to be careful on the timing of when to run the operation.

Agree on that 💯 

> I'm also finding it more difficult not working with ids/ordered uuids, that way we could base on a last seen id. Maybe we can do a list all with a timestamp (so list values created until X time, and also order by created).

Do you mean:
- DisableAll
- lastDisabledKey := Fetch the last disabled key ORDER BY created DESC
- List WHERE created <= lastDisabledKey LIMIT x OFFSET y

Reply:
> By lastDisabledKey you mean the disabled time of the last key disabled?

Indeed!